File: languages/go/client/record.go
package client

import (
	"context"
	"errors"

	"github.com/bloock/bloock-sdk-go/v2/entity/authenticity"
	"github.com/bloock/bloock-sdk-go/v2/entity/availability"
	"github.com/bloock/bloock-sdk-go/v2/entity/encryption"
	"github.com/bloock/bloock-sdk-go/v2/entity/record"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"
	"github.com/bloock/bloock-sdk-go/v2/internal/config"
)

// RecordClient provides functionality for creating records using various data sources and to interact with the [Bloock Record service].
//
// [Bloock Record service]: https://bloock.com
type RecordClient struct {
	bridgeClient bridge.BloockBridge
	configData   *proto.ConfigData
}

// NewRecordClient creates a new RecordClient with default configuration.
func NewRecordClient() RecordClient {
	return RecordClient{
		bridgeClient: bridge.NewBloockBridge(),
		configData:   config.NewConfigDataDefault(),
	}
}

// NewRecordClientWithConfig creates a new RecordClient with the provided configuration.
func NewRecordClientWithConfig(configData *proto.ConfigData) RecordClient {
	return RecordClient{
		bridgeClient: bridge.NewBloockBridge(),
		configData:   configData,
	}
}

// FromRecord creates a RecordBuilder from an existing record.
func (c RecordClient) FromRecord(record record.Record) RecordBuilder {
	return newRecordBuilder(record.ToProto(), proto.RecordTypes_RECORD, c.configData)
}

// FromLoader creates a RecordBuilder from a data loader.
func (c RecordClient) FromLoader(loader availability.Loader) RecordBuilder {
	return newRecordBuilder(loader.ToProto(), proto.RecordTypes_LOADER, c.configData)
}

// FromString creates a RecordBuilder from a string payload.
func (c RecordClient) FromString(str string) RecordBuilder {
	return newRecordBuilder(str, proto.RecordTypes_STRING, c.configData)
}

// FromHex creates a RecordBuilder from a hexadecimal string payload.
func (c RecordClient) FromHex(hex string) RecordBuilder {
	return newRecordBuilder(hex, proto.RecordTypes_HEX, c.configData)
}

// FromJSON creates a RecordBuilder from a JSON string payload.
func (c RecordClient) FromJSON(json string) RecordBuilder {
	return newRecordBuilder(json, proto.RecordTypes_JSON, c.configData)
}

// FromFile creates a RecordBuilder from a byte slice representing a file.
func (c RecordClient) FromFile(file_bytes []byte) RecordBuilder {
	return newRecordBuilder(file_bytes, proto.RecordTypes_FILE, c.configData)
}

// FromBytes creates a RecordBuilder from a byte slice payload.
func (c RecordClient) FromBytes(bytes []byte) RecordBuilder {
	return newRecordBuilder(bytes, proto.RecordTypes_BYTES, c.configData)
}

// RecordBuilder assists in constructing records with various configurations.
type RecordBuilder struct {
	payload     interface{}
	payloadType proto.RecordTypes
	configData  *proto.ConfigData
	signer      *proto.Signer
	encrypter   *proto.Encrypter
	decrypter   *proto.Encrypter
}

func newRecordBuilder(payload interface{}, payloadType proto.RecordTypes, configData *proto.ConfigData) RecordBuilder {
	return RecordBuilder{
		payload:     payload,
		payloadType: payloadType,
		configData:  configData,
	}
}

// WithSigner sets the signer for the RecordBuilder.
func (b RecordBuilder) WithSigner(signer authenticity.Signer) RecordBuilder {
	b.signer = signer.ToProto()
	return b
}

// WithEncrypter sets the encrypter for the RecordBuilder.
func (b RecordBuilder) WithEncrypter(encrypter encryption.Encrypter) RecordBuilder {
	b.encrypter = encrypter.ToProto()
	return b
}

// WithDecrypter sets the decrypter for the RecordBuilder.
func (b RecordBuilder) WithDecrypter(decrypter encryption.Encrypter) RecordBuilder {
	b.decrypter = decrypter.ToProto()
	return b
}

// Build constructs a record based on the RecordBuilder's configuration.
func (b RecordBuilder) Build() (record.Record, error) {
	bridgeClient := bridge.NewBloockBridge()

	var res *proto.RecordBuilderResponse
	var err error

	switch b.payloadType {
	case proto.RecordTypes_STRING:
		res, err = bridgeClient.Record().BuildRecordFromString(context.Background(), &proto.RecordBuilderFromStringRequest{
			ConfigData: b.configData,
			Payload:    b.payload.(string),
			Signer:     b.signer,
			Encrypter:  b.encrypter,
			Decrypter:  b.decrypter,
		})
	case proto.RecordTypes_BYTES:
		res, err = bridgeClient.Record().BuildRecordFromBytes(context.Background(), &proto.RecordBuilderFromBytesRequest{
			ConfigData: b.configData,
			Payload:    b.payload.([]byte),
			Signer:     b.signer,
			Encrypter:  b.encrypter,
			Decrypter:  b.decrypter,
		})
	case proto.RecordTypes_FILE:
		res, err = bridgeClient.Record().BuildRecordFromFile(context.Background(), &proto.RecordBuilderFromFileRequest{
			ConfigData: b.configData,
			Payload:    b.payload.([]byte),
			Signer:     b.signer,
			Encrypter:  b.encrypter,
			Decrypter:  b.decrypter,
		})
	case proto.RecordTypes_JSON:
		res, err = bridgeClient.Record().BuildRecordFromJson(context.Background(), &proto.RecordBuilderFromJSONRequest{
			ConfigData: b.configData,
			Payload:    b.payload.(string),
			Signer:     b.signer,
			Encrypter:  b.encrypter,
			Decrypter:  b.decrypter,
		})
	case proto.RecordTypes_HEX:
		res, err = bridgeClient.Record().BuildRecordFromHex(context.Background(), &proto.RecordBuilderFromHexRequest{
			ConfigData: b.configData,
			Payload:    b.payload.(string),
			Signer:     b.signer,
			Encrypter:  b.encrypter,
			Decrypter:  b.decrypter,
		})
	case proto.RecordTypes_RECORD:
		res, err = bridgeClient.Record().BuildRecordFromRecord(context.Background(), &proto.RecordBuilderFromRecordRequest{
			ConfigData: b.configData,
			Payload:    b.payload.(*proto.Record),
			Signer:     b.signer,
			Encrypter:  b.encrypter,
			Decrypter:  b.decrypter,
		})

	case proto.RecordTypes_LOADER:
		res, err = bridgeClient.Record().BuildRecordFromLoader(context.Background(), &proto.RecordBuilderFromLoaderRequest{
			ConfigData: b.configData,
			Loader:     b.payload.(*proto.Loader),
			Signer:     b.signer,
			Encrypter:  b.encrypter,
			Decrypter:  b.decrypter,
		})
	}

	if err != nil {
		return record.Record{}, err
	}

	if res.Error != nil {
		return record.Record{}, errors.New(res.Error.Message)
	}

	return record.NewRecordFromProto(res.Record, b.configData), nil
}

// GetDetails retrieves details about other Bloock services (Integrity, Authenticity, Encryption, Availability) 
// configured in the RecordBuilder.
func (b RecordBuilder) GetDetails() (record.RecordDetails, error) {
	bridgeClient := bridge.NewBloockBridge()

	res, err := bridgeClient.Record().GetDetails(context.Background(), &proto.GetDetailsRequest{
		ConfigData: b.configData,
		Payload:    b.payload.([]byte),
	})

	if err != nil {
		return record.RecordDetails{}, err
	}

	if res.Error != nil {
		return record.RecordDetails{}, errors.New(res.Error.Message)
	}

	return record.NewRecordDetailsFromProto(res.Details, b.configData), nil
}

File: languages/js/src/client/record.ts
import { BloockBridge } from "../bridge/bridge";
import {
  RecordBuilderFromStringRequest,
  RecordBuilderFromJSONRequest,
  RecordBuilderFromHexRequest,
  RecordBuilderFromBytesRequest,
  RecordBuilderFromFileRequest,
  RecordBuilderFromRecordRequest,
  RecordBuilderFromLoaderRequest,
  GetDetailsRequest
} from "../bridge/proto/record";
import { Record } from "../entity/record";
import { Signer } from "../entity/authenticity";
import { Encrypter } from "../entity/encryption";
import { Loader } from "../entity/availability";
import { NewConfigData } from "../config/config";
import { ConfigData } from "../bridge/proto/config";
import { RecordTypes } from "../bridge/proto/record_entities";
import { Signer as SignerProto } from "../bridge/proto/authenticity_entities";
import { Encrypter as EncrypterProto } from "../bridge/proto/encryption_entities";
import { RecordDetails } from "../entity/record/record-details";

/**
 * Provides functionality for creating records using various data sources and to interact with the [Bloock Record service](https://dashboard.bloock.com/login).
 */
export class RecordClient {
  private configData: ConfigData;

  /**
   * Creates a new RecordClient with default configuration.
   * @param configData 
   */
  constructor(configData?: ConfigData) {
    this.configData = NewConfigData(configData);
  }

  /**
   * Creates a RecordBuilder from a string payload.
   * @param str 
   * @returns 
   */
  public fromString(str: string): RecordBuilder {
    return new RecordBuilder(str, RecordTypes.STRING, this.configData);
  }

  /**
   * Creates a RecordBuilder from a JSON string payload.
   * @param json 
   * @returns 
   */
  public fromJson(json: any): RecordBuilder {
    return new RecordBuilder(
      JSON.stringify(json),
      RecordTypes.JSON,
      this.configData
    );
  }

  /**
   * Creates a RecordBuilder from a hexadecimal string payload.
   * @param hex 
   * @returns 
   */
  public fromHex(hex: string): RecordBuilder {
    return new RecordBuilder(hex, RecordTypes.HEX, this.configData);
  }

  /**
   * Creates a RecordBuilder from a byte slice payload.
   * @param bytes 
   * @returns 
   */
  public fromBytes(bytes: Uint8Array): RecordBuilder {
    return new RecordBuilder(bytes, RecordTypes.BYTES, this.configData);
  }

  /**
   * Creates a RecordBuilder from a byte slice representing a file.
   * @param bytes 
   * @returns 
   */
  public fromFile(bytes: Uint8Array): RecordBuilder {
    return new RecordBuilder(bytes, RecordTypes.FILE, this.configData);
  }

  /**
   * Creates a RecordBuilder from an existing record.
   * @param bytes 
   * @returns 
   */
  public fromRecord(bytes: Record): RecordBuilder {
    return new RecordBuilder(bytes, RecordTypes.RECORD, this.configData);
  }

  /**
   * Creates a RecordBuilder from a data loader.
   * @param loader 
   * @returns 
   */
  public fromLoader(loader: Loader): RecordBuilder {
    return new RecordBuilder(loader, RecordTypes.LOADER, this.configData);
  }
}

/**
 * Assists in constructing records with various configurations.
 */
export class RecordBuilder {
  payload: any;
  payloadType!: RecordTypes;
  signer: SignerProto | undefined;
  encrypter: EncrypterProto | undefined;
  decrypter: EncrypterProto | undefined;

  configData: ConfigData;

  /**
   * Creates a new RecordBuilder with default configuration.
   * @param payload 
   * @param payloadType 
   * @param configData 
   */
  constructor(payload: any, payloadType: RecordTypes, configData: ConfigData) {
    this.payload = payload;
    this.payloadType = payloadType;
    this.configData = configData;
  }

  /**
   * Sets the signer for the RecordBuilder.
   * @param signer 
   * @returns 
   */
  public withSigner(signer: Signer): RecordBuilder {
    this.signer = signer.toProto();
    return this;
  }

  /**
   * Sets the encrypter for the RecordBuilder.
   * @param encrypter 
   * @returns 
   */
  public withEncrypter(encrypter: Encrypter): RecordBuilder {
    this.encrypter = encrypter.toProto();
    return this;
  }

  /**
   * Sets the decrypter for the RecordBuilder.
   * @param decrypter 
   * @returns 
   */
  public withDecrypter(decrypter: Encrypter): RecordBuilder {
    this.decrypter = decrypter.toProto();
    return this;
  }

  /**
   * Constructs a record based on the RecordBuilder's configuration.
   * @returns 
   */
  async build(): Promise<Record> {
    const bridge = new BloockBridge();
    switch (this.payloadType) {
      case RecordTypes.STRING: {
        const req = RecordBuilderFromStringRequest.fromPartial({
          configData: this.configData,
          payload: this.payload,
          signer: this.signer,
          encrypter: this.encrypter,
          decrypter: this.decrypter
        });

        return bridge
          .getRecord()
          .BuildRecordFromString(req)
          .then(res => {
            if (res.error) {
              throw res.error;
            }
            return Record.fromProto(res.record!, this.configData);
          });
      }
      case RecordTypes.JSON: {
        const req = RecordBuilderFromJSONRequest.fromPartial({
          configData: this.configData,
          payload: this.payload,
          signer: this.signer,
          encrypter: this.encrypter,
          decrypter: this.decrypter
        });

        return bridge
          .getRecord()
          .BuildRecordFromJson(req)
          .then(res => {
            if (res.error) {
              throw res.error;
            }
            return Record.fromProto(res.record!, this.configData);
          });
      }
      case RecordTypes.HEX: {
        const req = RecordBuilderFromHexRequest.fromPartial({
          configData: this.configData,
          payload: this.payload,
          signer: this.signer,
          encrypter: this.encrypter,
          decrypter: this.decrypter
        });

        return bridge
          .getRecord()
          .BuildRecordFromHex(req)
          .then(res => {
            if (res.error) {
              throw res.error;
            }
            return Record.fromProto(res.record!, this.configData);
          });
      }
      case RecordTypes.BYTES: {
        const req = RecordBuilderFromBytesRequest.fromPartial({
          configData: this.configData,
          payload: this.payload,
          signer: this.signer,
          encrypter: this.encrypter,
          decrypter: this.decrypter
        });

        return bridge
          .getRecord()
          .BuildRecordFromBytes(req)
          .then(res => {
            if (res.error) {
              throw res.error;
            }
            return Record.fromProto(res.record!, this.configData);
          });
      }
      case RecordTypes.FILE: {
        const req = RecordBuilderFromFileRequest.fromPartial({
          configData: this.configData,
          payload: this.payload,
          signer: this.signer,
          encrypter: this.encrypter,
          decrypter: this.decrypter
        });

        return bridge
          .getRecord()
          .BuildRecordFromFile(req)
          .then(res => {
            if (res.error) {
              throw res.error;
            }
            return Record.fromProto(res.record!, this.configData);
          });
      }
      case RecordTypes.RECORD: {
        const req = RecordBuilderFromRecordRequest.fromPartial({
          configData: this.configData,
          payload: this.payload,
          signer: this.signer,
          encrypter: this.encrypter,
          decrypter: this.decrypter
        });

        return bridge
          .getRecord()
          .BuildRecordFromRecord(req)
          .then(res => {
            if (res.error) {
              throw res.error;
            }
            return Record.fromProto(res.record!, this.configData);
          });
      }
      case RecordTypes.LOADER: {
        const req = RecordBuilderFromLoaderRequest.fromPartial({
          configData: this.configData,
          loader: this.payload,
          signer: this.signer,
          encrypter: this.encrypter,
          decrypter: this.decrypter
        });

        return bridge
          .getRecord()
          .BuildRecordFromLoader(req)
          .then(res => {
            if (res.error) {
              throw res.error;
            }
            return Record.fromProto(res.record!, this.configData);
          });
      }
    }
    return Promise.reject(new Error("Unexpected record type"));
  }

  /**
   * Gets details about other Bloock services (Integrity, Authenticity, Encryption, Availability) configured in the RecordBuilder.
   * @returns 
   */
  async getDetails(): Promise<RecordDetails> {
    const bridge = new BloockBridge();
    const req = GetDetailsRequest.fromPartial({
      configData: this.configData,
      payload: this.payload
    });

    return bridge
      .getRecord()
      .GetDetails(req)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return RecordDetails.fromProto(res.details!);
      });
  }
}
